jerry-debugger/src/lib/protocol-handler.ts

```diff
@@ -0,0 +1,251 @@
+  private functionName?: Uint8Array;
+  private functions: FunctionMap = {};
+  private lastScriptID: number = 0;
+
+  constructor(delegate: JerryDebugProtocolDelegate) {
+    this.delegate = delegate;
+
+    this.byteConfig = {
+      cpointerSize: 0,
+      littleEndian: true,
+    };
+
+    this.functionMap = {
+      [SP.JERRY_DEBUGGER_CONFIGURATION]: this.onConfiguration,
+      [SP.JERRY_DEBUGGER_BYTE_CODE_CP]: this.onByteCodeCP,
+      [SP.JERRY_DEBUGGER_PARSE_FUNCTION]: this.onParseFunction,
+      [SP.JERRY_DEBUGGER_BREAKPOINT_LIST]: this.onBreakpointList,
+      [SP.JERRY_DEBUGGER_BREAKPOINT_OFFSET_LIST]: this.onBreakpointList,
+      [SP.JERRY_DEBUGGER_SOURCE_CODE]: this.onSourceCode,
+      [SP.JERRY_DEBUGGER_SOURCE_CODE_END]: this.onSourceCode,
+      [SP.JERRY_DEBUGGER_SOURCE_CODE_NAME]: this.onSourceCodeName,
+      [SP.JERRY_DEBUGGER_SOURCE_CODE_NAME_END]: this.onSourceCodeName,
+    };
+
+    this.stack = [{
+      isFunc: false,
+      line: 1,
+      column: 1,
+      name: '',
+      source: '',
+      lines: [],
+      offsets: [],
+    }];
+  }
+
+  // for the temporary mollification of lint
+  unused() {
+    this.maxMessageSize;
+  }
+
+  decodeMessage(format: string, message: Uint8Array, offset: number) {
+    return decodeMessage(this.byteConfig, format, message, offset);
+  }
+
+  onConfiguration(data: Uint8Array) {
+    console.log('[Configuration]');
+    if (data.length !== 5) {
+      this.abort('configuration message wrong size');
+      return;
+    }
+
+    this.maxMessageSize = data[1];
+    this.byteConfig.cpointerSize = data[2];
+    this.byteConfig.littleEndian = Boolean(data[3]);
+    this.version = data[4];
+
+    if (this.byteConfig.cpointerSize !== 2 && this.byteConfig.cpointerSize !== 4) {
+      this.abort('compressed pointer must be 2 or 4 bytes long');
+    }
+
+    if (this.version !== JERRY_DEBUGGER_VERSION) {
+      this.abort('incorrect target debugger version detected: ' + this.version
+                 + ' expected: ' + JERRY_DEBUGGER_VERSION);
+    }
+  }
+
+  onByteCodeCP(data: Uint8Array) {
+    console.log('[Byte Code CP]');
+    const frame = this.stack.pop();
+    if (!frame) {
+      throw new Error('missing parser stack frame');
+    }
+
+    const byteCodeCP = this.decodeMessage('C', data, 1)[0];
+    const func = new ParsedFunction(byteCodeCP, frame);
+
+    this.functions[byteCodeCP] = func;
+    if (this.stack.length > 0) {
+      return;
+    }
+
+    // FIXME: it seems like this is probably unnecessarily keeping the
+    //   whole file's source to this point?
+    func.source = this.source.split(/\r\n[\r\n]/);
+    func.sourceName = this.sourceName;
+  }
+
+  onParseFunction(data: Uint8Array) {
+    console.log('[Parse Function]');
+    const position = this.decodeMessage('II', data, 1);
+    this.stack.push({
+      isFunc: true,
+      line: position[0],
+      column: position[1],
+      name: cesu8ToString(this.functionName),
+      source: this.source,
```

Comment:
Later on, `this.source` is being split using `split(/\r\n[\r\n]/)`.
In `onSourceCode`, it's also being split.
I wonder if it makes sense to split it once and use the result everywhere instead (and ditch the non-splitted string.)